File: gesture_engine.py
"""
Gesture Engine – detects dynamic gestures from a rolling history of hand landmarks.

Supports: swipe (4 directions), pinch, and point gestures.
"""

import math
import time
from enum import Enum


class Gesture(Enum):
    NONE        = "None"
    SWIPE_LEFT  = "\u2190 Swipe Left"
    SWIPE_RIGHT = "Swipe Right \u2192"
    SWIPE_UP    = "\u2191 Swipe Up"
    SWIPE_DOWN  = "Swipe Down \u2193"
    PINCH       = "Pinch"
    POINT       = "Point"


class GestureEngine:
    """Analyses a rolling history of hand landmarks to detect dynamic gestures.

    Create one instance per tracked hand.
    """

    def __init__(
        self,
        swipe_threshold: float = 0.12,
        swipe_window: int = 18,
        pinch_threshold: float = 0.045,
    ):
        self.swipe_threshold = swipe_threshold
        self.swipe_window = swipe_window
        self.pinch_threshold = pinch_threshold

        self.gesture: Gesture = Gesture.NONE
        self.cursor: tuple[float, float] | None = None  # normalised (x, y)
        self.is_clicking: bool = False
        self.pinch_scale: float = 1.0

        self._history: list[dict] = []
        self._max_history = 25
        self._pinch_start_dist: float | None = None
        self._swipe_cooldown: int = 0

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def update(self, landmarks, fingers_extended: int) -> Gesture:
        """Feed the current frame's landmarks and finger count.

        `landmarks` is the MediaPipe NormalizedLandmarkList (list-like, 21 items).
        Returns the detected gesture.
        """
        index_tip = landmarks[8]
        thumb_tip = landmarks[4]

        self._history.append({
            "ix": index_tip.x, "iy": index_tip.y,
            "tx": thumb_tip.x, "ty": thumb_tip.y,
            "fingers": fingers_extended,
            "t": time.monotonic(),
        })
        if len(self._history) > self._max_history:
            self._history.pop(0)

        if self._swipe_cooldown > 0:
            self._swipe_cooldown -= 1

        thumb_index_dist = math.hypot(
            thumb_tip.x - index_tip.x,
            thumb_tip.y - index_tip.y,
        )

        # --- Priority: Pinch > Point > Swipe > None ---

        # 1. Pinch
        if thumb_index_dist < self.pinch_threshold:
            if self._pinch_start_dist is None:
                self._pinch_start_dist = max(thumb_index_dist, 0.001)
            self.pinch_scale = thumb_index_dist / self._pinch_start_dist

            if fingers_extended <= 2:
                self.gesture = Gesture.PINCH
                self.is_clicking = True
                self.cursor = (index_tip.x, index_tip.y)
            else:
                self.gesture = Gesture.PINCH
                self.is_clicking = False
                self.cursor = None
            return self.gesture

        # Pinch ended
        self._pinch_start_dist = None
        self.pinch_scale = 1.0
        self.is_clicking = False

        # 2. Point (only index extended)
        if fingers_extended == 1 and index_tip.y < landmarks[6].y:
            # MediaPipe: y increases downward, tip above PIP = extended
            self.gesture = Gesture.POINT
            self.cursor = (index_tip.x, index_tip.y)
            return self.gesture

        self.cursor = None

        # 3. Swipe (open hand, enough history, no cooldown)
        if (
            self._swipe_cooldown == 0
            and fingers_extended >= 3
            and len(self._history) >= self.swipe_window
        ):
            swipe = self._detect_swipe()
            if swipe is not None:
                self.gesture = swipe
                self._swipe_cooldown = self.swipe_window
                self._history.clear()
                return self.gesture

        # 4. Nothing
        self.gesture = Gesture.NONE
        return self.gesture

    def reset(self) -> None:
        self._history.clear()
        self.gesture = Gesture.NONE
        self.cursor = None
        self.is_clicking = False
        self.pinch_scale = 1.0
        self._pinch_start_dist = None
        self._swipe_cooldown = 0

    # ------------------------------------------------------------------
    # Swipe detection
    # ------------------------------------------------------------------

    def _detect_swipe(self) -> Gesture | None:
        window = self._history[-self.swipe_window:]
        first, last = window[0], window[-1]

        dx = last["ix"] - first["ix"]
        dy = last["iy"] - first["iy"]  # MediaPipe: y increases downward

        abs_dx = abs(dx)
        abs_dy = abs(dy)

        # Horizontal
        if abs_dx > self.swipe_threshold and abs_dx > abs_dy * 1.4:
            return Gesture.SWIPE_RIGHT if dx > 0 else Gesture.SWIPE_LEFT

        # Vertical (note: MediaPipe y is inverted vs screen "up")
        if abs_dy > self.swipe_threshold and abs_dy > abs_dx * 1.4:
            return Gesture.SWIPE_DOWN if dy > 0 else Gesture.SWIPE_UP

        return None

File: ios/HandGestureRecognizer/GestureEngine.swift
import Foundation
import Vision
import QuartzCore

// MARK: - Gesture types

enum HandGesture: String {
    case none       = "None"
    case swipeLeft  = "← Swipe Left"
    case swipeRight = "Swipe Right →"
    case swipeUp    = "↑ Swipe Up"
    case swipeDown  = "Swipe Down ↓"
    case pinch      = "Pinch"
    case point      = "Point"
}

// MARK: - Gesture Engine

/// Analyses a rolling history of hand landmarks to detect dynamic gestures.
/// One instance per hand slot.
final class GestureEngine: ObservableObject {

    // MARK: Published state

    @Published var gesture: HandGesture = .none
    @Published var cursorPosition: CGPoint?   // normalised, set when pointing
    @Published var isClicking: Bool = false    // pinch while pointing
    @Published var pinchScale: CGFloat = 1.0  // ratio vs pinch-start distance

    // MARK: Configuration

    /// Minimum normalised displacement for a swipe (0–1 range).
    var swipeThreshold: CGFloat = 0.12
    /// Maximum frames to evaluate for a swipe gesture.
    var swipeWindowSize: Int = 18
    /// Pinch triggers when thumb–index distance < this (normalised).
    var pinchDistanceThreshold: CGFloat = 0.045

    // MARK: Internal state

    private struct Snapshot {
        let indexTip: CGPoint
        let thumbTip: CGPoint
        let fingersExtended: Int
        let timestamp: Double
    }

    private var history: [Snapshot] = []
    private let maxHistory = 25

    /// Distance at the moment the pinch started (for scale calculation).
    private var pinchStartDistance: CGFloat?
    /// Cooldown: ignore swipes for this many frames after one fires.
    private var swipeCooldown: Int = 0

    // MARK: - Public API

    /// Call once per frame with the latest hand data.
    func update(with hand: HandData) {
        guard let indexTip = hand.landmarks[.indexTip],
              let thumbTip = hand.landmarks[.thumbTip] else {
            resetTransient()
            return
        }

        let snap = Snapshot(
            indexTip: indexTip,
            thumbTip: thumbTip,
            fingersExtended: hand.fingersExtended,
            timestamp: CACurrentMediaTime()
        )
        history.append(snap)
        if history.count > maxHistory { history.removeFirst() }

        // Decrement cooldown
        if swipeCooldown > 0 { swipeCooldown -= 1 }

        // --- Detection priority: Pinch > Point > Swipe > None ---

        let thumbIndexDist = distance(thumbTip, indexTip)

        // 1. Pinch
        if thumbIndexDist < pinchDistanceThreshold {
            if pinchStartDistance == nil {
                pinchStartDistance = max(thumbIndexDist, 0.001)
            }
            pinchScale = thumbIndexDist / pinchStartDistance!

            // If also pointing (1 finger), it's a click
            if hand.fingersExtended <= 2 {
                gesture = .pinch
                isClicking = true
                cursorPosition = indexTip
            } else {
                gesture = .pinch
                isClicking = false
                cursorPosition = nil
            }
            return
        }

        // Pinch ended
        pinchStartDistance = nil
        pinchScale = 1.0
        isClicking = false

        // 2. Point (only index finger extended)
        if hand.fingersExtended == 1,
           let indexPIP = hand.landmarks[.indexPIP],
           indexTip.y > indexPIP.y { // index is actually extended (Vision: y up)
            gesture = .point
            cursorPosition = indexTip
            return
        }

        cursorPosition = nil

        // 3. Swipe (need open hand: ≥3 fingers, enough history, no cooldown)
        if swipeCooldown == 0,
           hand.fingersExtended >= 3,
           history.count >= swipeWindowSize {
            if let swipe = detectSwipe() {
                gesture = swipe
                swipeCooldown = swipeWindowSize // prevent rapid re-fire
                // Clear history so the same motion isn't detected again
                history.removeAll()
                return
            }
        }

        // 4. Nothing
        gesture = .none
    }

    /// Call when the hand disappears.
    func reset() {
        history.removeAll()
        resetTransient()
        pinchStartDistance = nil
        swipeCooldown = 0
    }

    // MARK: - Swipe detection

    private func detectSwipe() -> HandGesture? {
        let window = Array(history.suffix(swipeWindowSize))
        guard let first = window.first, let last = window.last else { return nil }

        let dx = last.indexTip.x - first.indexTip.x
        let dy = last.indexTip.y - first.indexTip.y  // Vision: y increases upward

        let absDx = abs(dx)
        let absDy = abs(dy)

        // Horizontal swipe: dx dominant
        if absDx > swipeThreshold && absDx > absDy * 1.4 {
            return dx > 0 ? .swipeRight : .swipeLeft
        }

        // Vertical swipe: dy dominant
        if absDy > swipeThreshold && absDy > absDx * 1.4 {
            return dy > 0 ? .swipeUp : .swipeDown
        }

        return nil
    }

    // MARK: - Helpers

    private func resetTransient() {
        gesture = .none
        cursorPosition = nil
        isClicking = false
        pinchScale = 1.0
    }

    private func distance(_ a: CGPoint, _ b: CGPoint) -> CGFloat {
        hypot(a.x - b.x, a.y - b.y)
    }
}
